charms/tensorboard-controller/tests/integration/test_charm.py

```diff
@@ -0,0 +1,200 @@
+
+    await ops_test.model.wait_for_idle(
+        apps=[APP_NAME], status="active", raise_on_blocked=True, timeout=1000
+    )
+
+
+@tenacity.retry(
+    wait=tenacity.wait_exponential(multiplier=2, min=1, max=10),
+    stop=tenacity.stop_after_attempt(30),
+    reraise=True,
+)
+def assert_replicas(client, resource_class, resource_name, namespace):
+    """Test for replicas.
+
+    Retries multiple times to allow for tensorboard to be created (i.e. readyReplicas == 1).
+    """
+    tensorboard = client.get(resource_class, resource_name, namespace=namespace)
+    replicas = tensorboard.get("status", {}).get("readyReplicas")
+
+    resource_class_kind = resource_class.__name__
+    if replicas == 1:
+        logger.info(f"{resource_class_kind}/{resource_name} readyReplicas == {replicas}")
+    else:
+        logger.info(
+            f"{resource_class_kind}/{resource_name} readyReplicas == {replicas} (waiting for '1')"
+        )
+
+    assert replicas == 1, f"Waited too long for {resource_class_kind}/{resource_name}!"
+
+
+async def test_create_tensorboard(ops_test: OpsTest):
+    """Test Tensorboard creation."""
+    lightkube_client = Client()
+    load_in_cluster_generic_resources(lightkube_client)
+
+    # Create PVC for Tensorboard logs and Tensorboard
+    resources = codecs.load_all_yaml(
+        "\n---\n".join([PVC_TEMPLATE_FILE.read_text(), TENSORBOARD_TEMPLATE_FILE.read_text()]),
+        context={"pvc_name": PVC_NAME, "tensorboard_name": TENSORBOARD_NAME},
+    )
+    for rsc in resources:
+        lightkube_client.create(rsc, namespace=ops_test.model_name)
+
+    try:
+        tensorboard_ready = lightkube_client.get(
+            TENSORBOARD_RESOURCE,
+            name=TENSORBOARD_NAME,
+            namespace=ops_test.model_name,
+        )
+    except ApiError:
+        assert False
+    assert tensorboard_ready, f"Tensorboard {ops_test.model_name}/{TENSORBOARD_NAME} not found!"
+
+    assert_replicas(lightkube_client, TENSORBOARD_RESOURCE, TENSORBOARD_NAME, ops_test.model_name)
+
+
+@pytest.mark.abort_on_fail
+async def test_remove_with_resources_present(ops_test: OpsTest):
+    """Test remove with all resources deployed.
+
+    Verify that all deployed resources that need to be removed are removed.
+    """
+    # remove deployed charm and verify that it is deleted
+    await ops_test.model.remove_application(app_name=APP_NAME, block_until_done=True)
+    assert APP_NAME not in ops_test.model.applications
+
+    # verify that all resources that were deployed are deleted
+    lightkube_client = Client()
+
+    # verify that all created CRDs and Services in namespace are deleted
+    for rsc in [
+        CustomResourceDefinition,
+        Service,
+    ]:
+        rsc_list = lightkube_client.list(
+            rsc,
+            labels=[("app.juju.is/created-by", APP_NAME)],
+            namespace=ops_test.model_name,
+        )
+        assert not list(rsc_list)
```

Comment:
I think this label should also be on any Pod, PVC, or Tensorboard your charm creates, so you might be able to move them into this test

Reply:
Yeah that's true about the PVC.  And for the Pod, I guess we're not actually creating Pods but a StatefulSet or Deployment maybe?  I'd bet whatever resource is the parent of the Pod has the labels.

...but I wouldn't bet a ton, because sometimes this doesn't work how I expect :D